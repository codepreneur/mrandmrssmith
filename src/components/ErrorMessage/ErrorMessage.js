import React, { Fragment } from 'react'
import PropTypes from 'prop-types'
import clsx from 'clsx'

import get from 'lodash/fp/get'
import getOr from 'lodash/fp/getOr'

import { withStyles } from '@material-ui/core/styles'
import { interleaveTextWithComponent } from '../../utils/reactHelpers'

import Typography from '@material-ui/core/Typography'
import WarningIcon from '@material-ui/icons/Warning'

const styles = (theme) => {
  const unit = getOr(0, ['spacing', 'unit'], theme)
  const errorColor = get(['palette', 'error', 'main'], theme)
  return {
    container: {
      marginLeft: unit,
      marginRight: unit * 3,
      display: 'flex',
    },
    icon: {
      ...getOr({}, ['typography', 'h4'], theme),
      color: errorColor,
      marginRight: unit * 2,
    },
    refCode: {
      color: get(['palette', 'text', 'secondary'], theme),
    },
    refCodeGutter: {
      marginTop: '1em',
    },
    title: {
      color: errorColor,
    },
  }
}

const text = {
  title: 'We are sorry, something went wrong.',
  moreInfo: [
    'Please, refresh your browser and try again.',
    'If the problem persists, please contact tech support.',
  ],
  refCode: 'Reference code:',
}

const ErrorMessage = ({ refCode, textOnly, shortText, classes }) =>
  textOnly ? (
    <Fragment>
      <Typography
        variant="body2"
        component="span"
        gutterBottom={Boolean(refCode)}
      >
        <strong className={classes.title}>{text.title}</strong>
        {!shortText && (
          <Fragment>
            <br />
            {text.moreInfo.join(' ')}
          </Fragment>
        )}
      </Typography>
      {refCode && (
        <Typography
          className={classes.refCode}
          variant="caption"
          component="span"
        >
          {text.refCode} <strong>{refCode}</strong>
        </Typography>
      )}
    </Fragment>
  ) : (
    <section className={classes.container}>
      <WarningIcon className={classes.icon} />
      <div>
        <Typography variant="h6" gutterBottom>
          {text.title}
        </Typography>
        {!shortText && (
          <Typography variant="body2" gutterBottom={Boolean(refCode)}>
            {interleaveTextWithComponent('br', text.moreInfo)}
          </Typography>
        )}
        {refCode && (
          <Typography
            className={clsx(classes.refCode, classes.refCodeGutter)}
            variant="caption"
          >
            {text.refCode} <strong>{refCode}</strong>
          </Typography>
        )}
      </div>
    </section>
  )

ErrorMessage.propTypes = {
  classes: PropTypes.object.isRequired,
  refCode: PropTypes.string,
  shortText: PropTypes.bool,
  textOnly: PropTypes.bool,
}

export default withStyles(styles)(ErrorMessage)
