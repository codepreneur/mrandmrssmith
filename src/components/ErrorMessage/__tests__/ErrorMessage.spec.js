import React from 'react'
import { shallow } from 'enzyme'

import ErrorMessage from '../ErrorMessage'

describe('ErrorMessage component', () => {
  describe('smoke test', () => {
    it('should render without any errors', () => {
      expect(() => {
        shallow(<ErrorMessage />)
      }).not.toThrow()
    })

    it('should render correctly', () => {
      const wrapper = shallow(<ErrorMessage />)
      expect(wrapper).toMatchSnapshot()
      expect(wrapper.dive()).toMatchSnapshot()
    })
  })

  describe('refCode test', () => {
    it('should not display ref code section if no refCode passed', () => {
      expect(
        shallow(<ErrorMessage />)
          .dive()
          .exists('.refCode'),
      ).toBe(false)
    })

    it('should not add bottom gutter if no refCode passed', () => {
      expect(
        shallow(<ErrorMessage />)
          .dive()
          .findWhere(
            (node) =>
              node.prop('variant') === 'body2' && !node.prop('gutterBottom'),
          ),
      ).toHaveLength(1)
    })

    it('should display ref code section if refCode passed', () => {
      expect(
        shallow(<ErrorMessage refCode="REF_CODE_XYZ" />)
          .dive()
          .exists('.refCode'),
      ).toBe(true)
    })

    it('should add bottom gutter if refCode passed', () => {
      expect(
        shallow(<ErrorMessage refCode="REF_CODE_XYZ" />)
          .dive()
          .findWhere(
            (node) =>
              node.prop('variant') === 'body2' && node.prop('gutterBottom'),
          ),
      ).toHaveLength(1)
    })

    it('should display correct ref', () => {
      expect(
        shallow(<ErrorMessage refCode="REF_CODE_XYZ" />)
          .dive()
          .find('.refCode')
          .find('strong')
          .text(),
      ).toBe('REF_CODE_XYZ')
    })
  })

  describe('textOnly test', () => {
    it('should display container without textOnly', () => {
      expect(
        shallow(<ErrorMessage />)
          .dive()
          .hasClass('container'),
      ).toBe(true)
    })

    it('should display without container when textOnly', () => {
      expect(
        shallow(<ErrorMessage textOnly />)
          .dive()
          .hasClass('container'),
      ).toBe(false)
    })
  })

  describe('shortText test', () => {
    it('should display long variant', () => {
      const wrapper = shallow(<ErrorMessage />)
      expect(
        wrapper.dive().findWhere((node) => node.prop('variant') === 'h6'),
      ).toHaveLength(1)
      expect(
        wrapper.dive().findWhere((node) => node.prop('variant') === 'body2'),
      ).toHaveLength(1)
    })

    it('should display long variant', () => {
      const wrapper = shallow(<ErrorMessage shortText />)
      expect(
        wrapper.dive().findWhere((node) => node.prop('variant') === 'h6'),
      ).toHaveLength(1)
      expect(
        wrapper.dive().findWhere((node) => node.prop('variant') === 'body2'),
      ).toHaveLength(0)
    })

    it('should display long variant combined with textOnly', () => {
      expect(
        shallow(<ErrorMessage textOnly />)
          .dive()
          .findWhere((node) => node.prop('variant') === 'body2')
          .first()
          .children(),
      ).toHaveLength(3)
    })

    it('should display short variant combined with textOnly', () => {
      expect(
        shallow(<ErrorMessage shortText textOnly />)
          .dive()
          .findWhere((node) => node.prop('variant') === 'body2')
          .first()
          .children(),
      ).toHaveLength(1)
    })
  })
})
