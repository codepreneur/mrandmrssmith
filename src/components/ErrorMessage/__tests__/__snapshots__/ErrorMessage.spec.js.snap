// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorMessage component smoke test should render correctly 1`] = `
<ErrorMessage
  classes={null}
/>
`;

exports[`ErrorMessage component smoke test should render correctly 2`] = `
<section
  className="container"
>
  <WarningIcon
    className="icon"
  />
  <div>
    <WithStyles(ForwardRef(Typography))
      gutterBottom={true}
      variant="h6"
    >
      We are sorry, something went wrong.
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      gutterBottom={false}
      variant="body2"
    >
      <span
        key="outer1"
      >
        Please, refresh your browser and try again.
      </span>
      <br
        key="inner2"
      />
      <span
        key="outer3"
      >
        If the problem persists, please contact tech support.
      </span>
    </WithStyles(ForwardRef(Typography))>
  </div>
</section>
`;
